File: AgentLens/Services/Cast/CastDevice.swift
import Foundation

// MARK: - Cast Device
//
// Concrete representation of a Cast-capable device discovered on the LAN.
// We deduplicate on `serviceName` (the underlying mDNS instance name) and
// fall back to host:port for the equality contract.

public struct CastDevice: Hashable, Sendable, Codable {

    /// Stable mDNS service name, e.g. `Google-Nest-Hub-dec04a601c00269a3...`.
    /// We use this as the canonical identifier; it survives IP changes.
    public let serviceName: String

    /// Friendly name shown in the wizard UI ("Living Room Hub").
    public let friendlyName: String

    /// Last-known LAN IP. Refreshed whenever discovery picks the device up
    /// again — Cast devices frequently move between IPs on DHCP renewal.
    public var host: String

    /// Cast TLS port. Always 8009 in practice but the protocol allows
    /// device-side override, so we keep it flexible.
    public var port: Int

    /// Hardware model (`Google Nest Hub`, `Chromecast`, etc.) sourced from
    /// the `md` mDNS TXT record.
    public let model: String

    /// `id` field from the TXT record — UUID assigned by Google to the
    /// device at first setup; doubles as a human-stable ID.
    public let identifier: String

    public var lastSeenAt: Date

    /// Whether this device is likely to render web content. Audio-only
    /// devices (Nest Mini, Nest Audio, Google Home, Home Max) return
    /// `NOT_FOUND` when DashCast tries to launch on them. Set by
    /// `CastDiscovery` based on the Cast capability bitmask + model
    /// name heuristics. Defaults to `true` for safety — false positives
    /// here would silently hide a Nest Hub.
    public var supportsDisplay: Bool

    public init(
        serviceName: String,
        friendlyName: String,
        host: String,
        port: Int,
        model: String,
        identifier: String,
        lastSeenAt: Date = Date(),
        supportsDisplay: Bool = true
    ) {
        self.serviceName = serviceName
        self.friendlyName = friendlyName
        self.host = host
        self.port = port
        self.model = model
        self.identifier = identifier
        self.lastSeenAt = lastSeenAt
        self.supportsDisplay = supportsDisplay
    }

    /// Convenience for UI grouping — collapses `Google-Nest-Hub-Max-…` /
    /// `Chromecast-…` to icon-friendly buckets.
    public var iconKind: IconKind {
        let lower = model.lowercased()
        if lower.contains("nest hub max") { return .nestHubMax }
        if lower.contains("nest hub") { return .nestHub }
        if lower.contains("chromecast") { return .chromecast }
        if lower.contains("nest mini") || lower.contains("nest audio") { return .nestSpeaker }
        return .generic
    }

    public enum IconKind: String, Sendable, Codable {
        case nestHub
        case nestHubMax
        case chromecast
        case nestSpeaker
        case generic
    }
}

File: windows/integrations/cast/OpenBurnBar.Integrations.Cast/Model/CastDevice.cs
// Parity source: AgentLens/Services/Cast/CastDevice.swift (struct CastDevice)

using System;

namespace OpenBurnBar.Integrations.Cast.Model;

/// <summary>
/// Icon-friendly bucket for a Cast device, collapsing model strings into a
/// small set of UI glyph categories. Port of Swift <c>CastDevice.IconKind</c>.
/// </summary>
public enum CastIconKind
{
    /// <summary>A Google Nest Hub (7").</summary>
    NestHub,

    /// <summary>A Google Nest Hub Max (10").</summary>
    NestHubMax,

    /// <summary>A Chromecast dongle / Chromecast with Google TV.</summary>
    Chromecast,

    /// <summary>An audio-only Nest speaker (Nest Mini / Nest Audio).</summary>
    NestSpeaker,

    /// <summary>Anything else that announces <c>_googlecast._tcp</c>.</summary>
    Generic,
}

/// <summary>
/// Concrete representation of a Cast-capable device discovered on the LAN.
/// Deduplicated on <see cref="ServiceName"/> (the mDNS instance name), which
/// survives IP changes. Port of Swift <c>CastDevice</c> (Hashable/Codable value
/// type) — modeled here as an immutable record with value equality.
/// </summary>
public sealed record CastDevice
{
    /// <summary>
    /// Stable mDNS service name, e.g.
    /// <c>Google-Nest-Hub-dec04a601c00269a3...</c>. Canonical identity; survives
    /// IP changes.
    /// </summary>
    public required string ServiceName { get; init; }

    /// <summary>Friendly name shown in the wizard UI ("Living Room Hub").</summary>
    public required string FriendlyName { get; init; }

    /// <summary>Last-known LAN IP; refreshed on every re-discovery (DHCP churn).</summary>
    public required string Host { get; init; }

    /// <summary>Cast TLS port — 8009 in practice, device-overridable by protocol.</summary>
    public required int Port { get; init; }

    /// <summary>Hardware model from the <c>md</c> TXT record.</summary>
    public required string Model { get; init; }

    /// <summary>The <c>id</c> TXT field — Google-assigned per-device UUID.</summary>
    public required string Identifier { get; init; }

    /// <summary>Last time discovery observed this device (UTC).</summary>
    public DateTimeOffset LastSeenAt { get; init; } = DateTimeOffset.UtcNow;

    /// <summary>
    /// Whether this device likely renders web content. Audio-only devices return
    /// <c>NOT_FOUND</c> when DashCast tries to launch; defaults to
    /// <see langword="true"/> so a false negative never silently hides a Hub.
    /// </summary>
    public bool SupportsDisplay { get; init; } = true;

    /// <summary>Convenience UI bucket — collapses the model string to an icon kind.</summary>
    public CastIconKind IconKind
    {
        get
        {
            var lower = Model.ToLowerInvariant();
            if (lower.Contains("nest hub max"))
            {
                return CastIconKind.NestHubMax;
            }

            if (lower.Contains("nest hub"))
            {
                return CastIconKind.NestHub;
            }

            if (lower.Contains("chromecast"))
            {
                return CastIconKind.Chromecast;
            }

            if (lower.Contains("nest mini") || lower.Contains("nest audio"))
            {
                return CastIconKind.NestSpeaker;
            }

            return CastIconKind.Generic;
        }
    }
}
